### PR TITLE
Added support for passive server

### PR DESCRIPTION
**Pull request by amoralej**
Currently, dlrn infra in RDO project is formed by two servers a
primary and a passive one. This commit adds proper support for both
servers to use the same hiera config file by creating a new parameter
server_type which can be set to primary (default) or passive.

Even when enabled in workers hiera configuration, passive servers will not enable:
- cron jobs for run-dlrn.sh
- rsync of new builds (they are suposed to be the destination of the synchronization)
- Opening reviews in gerrit automatically on FTBFS
- Sending mails on build failure
Additionally, passive servers will configure apache to redirect to buildlogs when users
trying to access to current-passed-ci or current-tripleo

As part of this commit, apache configuration is fully integrated in puppet-dlrn and
no more ssl_setup.sh is required. In case certificates are not installed, parameter
enable_https con be set to false.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -86,6 +86,17 @@ The DLRN instance architecture includes a secondary server, where all repos are
...
+####`server_type`
+Defines the server_type. It can be set to primary (default) or passive. Even when enabled in workers hiera configuration, passive servers will not enable:
+- cron jobs for run-dlrn.sh
+- rsync of new builds (they are suposed to be the destination of the synchronization)
+- Opening reviews in gerrit automatically on FTBFS
+- Sending mails on build failure
+Additionally, passive servers will redirect to buildlogs when users trying to access to current-passed-ci or current-tripleo
+
+####`enable_https`
+Enable https in apache configuration (requires proper certificates installed in the system). Default is true.
```
Comment (javierpena) [inline README.md]: imho, default should be false. We are requiring some specific certificates, which may or may not be available during the initial execution, so it is safest to not enable HTTPS by default.

File: README.md [hunk 2]
```diff
@@ -214,12 +225,15 @@ This is a user to run Gerrit reviews for packages after build failures. If set t
 ####`gerrit_email` 
 This is the email for the user to run Gerrit reviews for packages after build failures. It is required when `gerrit_user` is set, and ignored otherwise.
 
-###`rsyncdest`
+####`rsyncdest`
```
Comment (javierpena) [inline README.md]: Good catch!

File: examples/passive.pp [hunk 3]
```diff
@@ -0,0 +1,8 @@
+class { 'dlrn': 
+#  backup_server          => 'testbackup.example.com',
```
Comment (javierpena) [inline examples/passive.pp]: If this is a passive site, let's remove the backup_server commented option.

File: examples/passive.pp [hunk 4]
```diff
@@ -0,0 +1,8 @@
...
+  sshd_port              => 3300,
+  mock_tmpfs_enable      => false,
+  server_type            => 'passive',
+#  enable_https           => true,
```
Comment (javierpena) [inline examples/passive.pp]: As said before, let's uncomment and set enable_https to false by default.

File: examples/site.pp [hunk 5]
```diff
@@ -2,5 +2,7 @@
 #  backup_server          => 'testbackup.example.com',
   sshd_port              => 3300,
   mock_tmpfs_enable      => false,
+#  server_type            => 'primary',
```
Comment (javierpena) [inline examples/site.pp]: Let's be explicit and uncomment.

File: manifests/common.pp [hunk 6]
```diff
@@ -88,6 +82,11 @@
     service => 'http',
     zone    => 'public',
   } ->
+  firewalld_service { 'Allow HTTPS':
```
Comment (javierpena) [inline manifests/common.pp]: This should only be done if ::dlrn::enable_https is set to true

File: manifests/init.pp [hunk 7]
```diff
@@ -30,6 +43,8 @@
   $sshd_port         = 3300,
   $backup_server     = undef,
   $mock_tmpfs_enable = false,
+  $server_type       = "primary",
+  $enable_https      = true,
```
Comment (javierpena) [inline manifests/init.pp]: As said before, let's default to false.

File: manifests/web.pp [hunk 8]
```diff
@@ -4,8 +4,56 @@
 #
 
 class dlrn::web(
+  $enable_https = $dlrn::enable_https,
```
Comment (javierpena) [inline manifests/web.pp]: Is this parameter documented in README.md?

File: manifests/web.pp [hunk 9]
```diff
@@ -4,8 +4,56 @@
...
 ){
 
+  if $enable_https {
+    $cert_file  = '/etc/pki/tls/certs/trunk_rdoproject_org.crt'
+    $cert_key   = '/etc/pki/tls/private/trunk.rdoproject.org.key'
+    $cert_chain = '/etc/pki/tls/certs/DigiCertCA.crt'
```
Comment (javierpena) [inline manifests/web.pp]: I don't like the idea of hardcoding these variables. What if we set them as variables in the hiera file?

File: manifests/worker.pp [hunk 10]
```diff
@@ -89,7 +89,8 @@
   $gerrit_user    = undef,
   $gerrit_email   = undef,
   $rsyncdest      = undef,
-  $rsyncport      = 22) {
+  $rsyncport      = 22 ,
+  $server_type    = $dlrn::server_type ) {
```
Comment (javierpena) [inline manifests/worker.pp]: Also, not documented (I think)

File: manifests/worker.pp [hunk 11]
```diff
@@ -206,14 +207,14 @@
   }
 
   cron { "${name}-logrotate":
-    command => "/usr/bin/find /home/${name}/dlrn-logs/*.log -mtime +2 -exec rm {} \;",
+    command => "/usr/bin/find /home/${name}/dlrn-logs/*.log -mtime +2 -exec rm {} \\;",
```
Comment (javierpena) [inline manifests/worker.pp]: Have you tested this? I'm pretty sure it had to be \; instead of \\;

File: manifests/worker.pp [hunk 12]
```diff
@@ -271,8 +272,20 @@
     }
   }
 
+  # Add compatibility redirects
+  if $name =~ /^(centos)\-(.*)/ and $server_type == "passive"  {
```
Comment (javierpena) [inline manifests/worker.pp]: We should use single quotes for consistency, any time there is no variable substitution inside the string.

File: spec/defines/dlrn_worker_spec.rb [hunk 13]
```diff
@@ -348,7 +353,98 @@
...
         .with_content(/rsyncport=1022$/)
     end
+  end
+
+  context 'with parameter server_type = passive' do
+    before :each do
+      params.merge!(:server_type    => 'passive')
+    end
+ 
+    context 'with centos-mitacka name' do
```
Comment (javierpena) [inline spec/defines/dlrn_worker_spec.rb]: Typo

**Comment by javierpena** [comment]
Thanks! I like the idea of puppetizing the https configuration, but I think we should change the default (see comments inline). The rest are small details here and there.

File: manifests/worker.pp (hunk 11 above)
Comment (amoralej) [inline manifests/worker.pp]: Yes, it's tested. Without the double \ , rake validate fails although the cron job is ok. With double backslash, the actual cron job is correctly created:

0 4 * * * /usr/bin/find /home/centos-master/dlrn-logs/*.log -mtime +2 -exec rm {} \;

(taken from an actual puppet apply.

Comment (javierpena) [inline manifests/worker.pp]: Ok then, thanks!

**Comment by amoralej** [comment]
I've added some fixes accordint to your feedback on first commit.

**Comment by javierpena** [comment]
This version looks mostly ok to me. Only a couple things to fix:

- In web.pp, make apache::vhost { 'ssl-trunk.rdoproject.org'} depend on the 3 file resources, so the ssl vhost is only created when the certs are available.
- rake lint has some complains.

**Comment by amoralej** [comment]
I'm going to close this PR, and split into two different ones as it's becoming too big for just one change.

**Pull request closed without merging**